### PR TITLE
8mm evk a53 support

### PR DESCRIPTION
**Pull request by MrVan**
This goal is to support Zephyr runs on top of Jailhouse hypervisor. 

This patchset is the first step to make Zephyr runs on i.MX8MM A53 core from U-Boot.

### CODE REVIEW

File: boards/arm/mimx8mm_evk_a53/board.cmake [hunk 1]
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (hakehuang) [inline boards/arm/mimx8mm_evk_a53/board.cmake:1]: need add debugger support
`
board_set_debugger_ifnset(jlink)
board_set_flasher_ifnset(jlink)

board_runner_args(jlink "--device=MIMX8MD6_A53")
include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
`

**Review by hakehuang — COMMENTED** [review]
suggest to create an issues to explain on how to use this with uboot + jailhouse.

File: boards/arm/mimx8mm_evk_a53/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+NXP i.MX8MM EVK (Cortex-A53)
+#################################
+
+Overview
+********
+
+i.MX8M Mini LPDDR4 EVK board is based on NXP i.MX8M Mini applications
+processor, composed of a quad Cortex®-A53 cluster and a single Cortex®-M4 core.
+Zephyr OS is ported to run on the Cortex®-M4 core.
```
Comment (carlescufi) [inline boards/arm/mimx8mm_evk_a53/doc/index.rst:11]: Did you mean to say that Zephyr already runs on the M4, but now it can also run on the A53? It's a bit confusing as it is written now.

File: soc/arm/nxp_imx/mcimx8mm_a53/Kconfig.defconfig.mimx8mm_a53 [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2020 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_MIMX8MM_A53
+
+config SOC
+	default "mimx8mm6"
+
+config FLASH_SIZE
```
Comment (galak) [inline soc/arm/nxp_imx/mcimx8mm_a53/Kconfig.defconfig.mimx8mm_a53:9]: Would be good for those to be coming in via devicetree and having `zephyr,flash` set.

**Comment by Mani-Sadhasivam** [comment]
@MrVan FWIW, I worked on this a while ago but never managed to clean it up and submit the PR:
https://github.com/Mani-Sadhasivam/zephyr/commits/jailhouse

I had to modify the GIC driver to support GIC v2.

File: boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+CONFIG_BOARD_MIMX8MM_EVK_A53=y
+CONFIG_ARM64_VA_BITS_36=y
+CONFIG_ARM64_PA_BITS_36=y
+CONFIG_ARM_ARCH_TIMER=y
+
+# Zephyr Kernel Configuration
+CONFIG_SRAM_BASE_ADDRESS=0x40000000
+CONFIG_XIP=n
+CONFIG_SRAM_SIZE=1024
+CONFIG_ARM64_PA_BITS_48=y
```
Comment (carlocaione) [inline boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:15]: you have `CONFIG_ARM64_PA_BITS_36=y` at line 8 up here.

File: boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+# Platform Configuration
+CONFIG_SOC_SERIES_IMX8MM_A53=y
+CONFIG_SOC_MIMX8MM_A53=y
+CONFIG_BOARD_MIMX8MM_EVK_A53=y
+CONFIG_ARM64_VA_BITS_36=y
+CONFIG_ARM64_PA_BITS_36=y
+CONFIG_ARM_ARCH_TIMER=y
+
+# Zephyr Kernel Configuration
+CONFIG_SRAM_BASE_ADDRESS=0x40000000
```
Comment (carlocaione) [inline boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:12]: This should not be necessary because the value is directly read from the DTS.

File: boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+CONFIG_XIP=n
+CONFIG_SRAM_SIZE=1024
+CONFIG_ARM64_PA_BITS_48=y
+
+# Serial Drivers
+CONFIG_SERIAL=y
```
Comment (carlocaione) [inline boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:18]: you have `CONFIG_SERIAL=y` but no serial driver enabled?

File: soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+#include <arch/arm/aarch64/arm_mmu.h>
+
+#define SZ_1K	1024
+
+static const struct arm_mmu_region mmu_regions[] = {
+
+	MMU_REGION_FLAT_ENTRY("GIC",
+			      DT_REG_ADDR_BY_IDX(DT_INST(0, arm_gic), 0),
+			      DT_REG_SIZE_BY_IDX(DT_INST(0, arm_gic), 0),
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
```
Comment (carlocaione) [inline soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:17]: instead of `MT_RW` please use `MT_P_RW_U_RW` or similar.

File: soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+
+	MMU_REGION_FLAT_ENTRY("ANA_PLL",
+			      DT_REG_ADDR(DT_INST(0, nxp_imx_ana)),
+			      DT_REG_SIZE(DT_INST(0, nxp_imx_ana)),
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("DRAM",
+			      CONFIG_SRAM_BASE_ADDRESS,
+			      CONFIG_SRAM_SIZE * SZ_1K,
```
Comment (carlocaione) [inline soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:42]: Use `KB()` instead of `* SZ_1K`

File: soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
+
+
+	MMU_REGION_FLAT_ENTRY("ANA_PLL",
+			      DT_REG_ADDR(DT_INST(0, nxp_imx_ana)),
+			      DT_REG_SIZE(DT_INST(0, nxp_imx_ana)),
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("DRAM",
+			      CONFIG_SRAM_BASE_ADDRESS,
```
Comment (carlocaione) [inline soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:41]: DRAM vs SRAM. But why do you need this? In general it should not be needed to map the whole SRAM.

File: boards/arm/mimx8mm_evk_a53/doc/index.rst (hunk 2 above)
Comment (MrVan) [reply boards/arm/mimx8mm_evk_a53/doc/index.rst:11]: It runs on A53 now.

**Comment by MrVan** [comment]
> @MrVan FWIW, I worked on this a while ago but never managed to clean it up and submit the PR:
> https://github.com/Mani-Sadhasivam/zephyr/commits/jailhouse
> 
> I had to modify the GIC driver to support GIC v2.

Wow. what usecases you are making zephyr on Jailhouse for?

File: soc/arm/nxp_imx/mcimx8mm_a53/Kconfig.defconfig.mimx8mm_a53 (hunk 3 above)
Comment (MrVan) [reply soc/arm/nxp_imx/mcimx8mm_a53/Kconfig.defconfig.mimx8mm_a53:9]: Currently I not have the flash on board, so use fixed value here should be ok? I could update to use your suggested method when we use flash in future.

File: boards/arm/mimx8mm_evk_a53/board.cmake (hunk 1 above)
Comment (MrVan) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: I am not sure, but I not see this "./boards/arm/bcm958402m2_a72/board.cmake" and "boards/arm/qemu_cortex_a53/board.cmake",  Is it really needed?

File: soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c (hunk 9 above)
Comment (MrVan) [reply soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:41]: I am thinking we need memory for IPC usage, between zephyr and Linux.

File: soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c (hunk 7 above)
Comment (MrVan) [reply soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:17]: fix in next version

File: boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig (hunk 6 above)
Comment (MrVan) [reply boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:18]: I have this in Kconfig

`
if SERIAL

config UART_MCUX_IUART
        default y

endif # SERIAL
`

File: soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c (hunk 9 above)
Comment (carlocaione) [reply soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:41]: Ok, but mapping the whole SRAM is probably an overkill. Can you define a memory region to be used for IPC and map that instead?

File: boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig (hunk 4 above)
Comment (MrVan) [reply boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:12]: Resolved.

File: soc/arm/nxp_imx/mcimx8mm_a53/Kconfig.defconfig.mimx8mm_a53 (hunk 3 above)
Comment (MrVan) [reply soc/arm/nxp_imx/mcimx8mm_a53/Kconfig.defconfig.mimx8mm_a53:12]: Changed to use device tree.

File: soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c (hunk 9 above)
Comment (MrVan) [reply soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c]: sure. Currently I have not add that piece memory, will add that in future PRs.

**Comment by MrVan** [comment]
updated, please help check.

**Comment by MrVan** [comment]
Updated, please help check

File: drivers/clock_control/clock_control_mcux_ccm.c [hunk 10]
```diff
@@ -103,11 +104,31 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
...
+		} else if (mux == 1) {
+			*rate = CLOCK_GetPllFreq(kCLOCK_SystemPll1Ctrl) /
+					(CLOCK_GetRootPreDivider(kCLOCK_RootUart4)) /
+					(CLOCK_GetRootPostDivider(kCLOCK_RootUart4)) /
+					10;
+		}
+		break;
+
+	case IMX_CCM_UART2_CLK:
+
```
Comment (carlocaione) [inline drivers/clock_control/clock_control_mcux_ccm.c:121]: nit: empty line

File: dts/arm/nxp/nxp_imx8mm_a53.dtsi [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 14 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 11 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>,
+			     <GIC_PPI 10 IRQ_TYPE_LEVEL
+			      IRQ_DEFAULT_PRIORITY>;
+		label = "arch_timer";
+		interrupt-parent = <&gic>;
+	};
```
Comment (carlocaione) [inline dts/arm/nxp/nxp_imx8mm_a53.dtsi:45]: nit: add empty line here

File: dts/arm/nxp/nxp_imx8mm_a53.dtsi [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+		zephyr,console = &uart2;
+		zephyr,shell-uart = &uart2;
+		zephyr,sram = &sram0;
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
```
Comment (carlocaione) [inline dts/arm/nxp/nxp_imx8mm_a53.dtsi:26]: given that we have `CONFIG_MP_NUM_CPUS=2` should we have `cpu@1` here as well?

**Review by carlocaione — APPROVED** [review]

**Comment by MrVan** [comment]
@MaureenHelm for the build failure 

> /var/lib/buildkite-agent/zephyr-module-cache/modules/hal/nxp/mcux/devices/MIMX8MM6/MIMX8MM6_cm4.h:235:10: fatal error: core_cm4.h: No such file or directory
--
  | 235 \| #include "core_cm4.h"                  /* Core Peripheral Access Layer */
  | \|          ^~~~~~~~~~~~

Should I send a PR to nxp hal?

File: boards/arm/mimx8mm_evk_a53/board.cmake (hunk 1 above)
Comment (hakehuang) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: without that you can not use west flash to download the application. how do you program the board now?

Comment (MrVan) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: Use U-Boot to load/kick the zephyr.bin or use Linux to load it for hypervisor usage. zephyr.bin is stored in sdcard fat partition.

Comment (hakehuang) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: @MrVan  the official way is to use `west flash`

Comment (carlocaione) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: Nope, using U-Boot is fine and `west flash` is totally optional here. Not always you want to use the debugger to load the image also considering that you do not always have a debugger.

Comment (hakehuang) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: then user need extra steps to setup uboot, and as uboot is evovled overtime, this needs some extra documents and efforts.

Comment (carlocaione) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: How to use U-Boot is already documented in this commit and that is enough. Fine to me if we can use `west flash` to flash the board as well but there is no such a thing as "official" way to flash or use the board. @MrVan is this a problem adding support for JLink?

Comment (MrVan) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: I am not familar with JLink. To me, we have sdcard to hold zephyr.bin, zephyr.bin is copied to the sdcard by mounting the sdcard to Linux PC. Could west flash support writing fat partition of a sdcard? Actually we have yocto Linux for image generating.

Comment (hakehuang) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: @MrVan west does not support fat partition write, it is a different way. It may not be officially defined to use west flash, but basically we expect to have an easy to use solution.

Comment (carlocaione) [reply boards/arm/mimx8mm_evk_a53/board.cmake:1]: > @MrVan west does not support fat partition write, it is a different way. It may not be officially defined to use west flash, but basically we expect to have an easy to use solution. Besides if there no debuger tools supported, how users suppose to debug the program? IMHO, it is not a good enough enablement.

I disagree. U-Boot + SD is indeed an easy solution to use in this case (and pretty standard when dealing with AArch64 boards, much more common than having a J-Link available) and you cannot expect that everybody has access to a debugger / J-Link or that every board has a debug port easily accessible. So, this enablement is good enough for merging.

Now, that said @MrVan, can you add what suggested here https://github.com/zephyrproject-rtos/zephyr/pull/32694#discussion_r583531459 so that `west flash` can be used when a debugger is available? This doesn't prevent the user from using U-Boot and the method you already described.

Please, also update the documentation describing both the methods.

**Review by MaureenHelm — APPROVED** [review]
> @MaureenHelm for the build failure
> 
> > /var/lib/buildkite-agent/zephyr-module-cache/modules/hal/nxp/mcux/devices/MIMX8MM6/MIMX8MM6_cm4.h:235:10: fatal error: core_cm4.h: No such file or directory
> > --
> >   | 235 | #include "core_cm4.h"                  /* Core Peripheral Access Layer */
> >   | |          ^~~~~~~~~~~~
> 
> Should I send a PR to nxp hal?

Yes. Please also send a PR to https://github.com/NXPmicro/mcux-sdk

**Comment by MrVan** [comment]
NXP HAL PR: https://github.com/zephyrproject-rtos/hal_nxp/pull/80

Add a few more commits to address build warning for ARM64 and support Jailhouse hypervisor.

**Comment by MrVan** [comment]
https://github.com/NXPmicro/mcux-sdk/pull/11 created.

**Comment by MrVan** [comment]
Finally find some time to rework to drop NXP HAL MCU dependency, because this is for Cortex-A usage, and NXP not has Cortex-A HAL SDK support, so added a new uart driver for i.MX8MM A53 usage without depends on NXP HAL MCU SDK. Hope this could make the PR in, thanks.

**Review by hakehuang — APPROVED** [review]

**Comment by MrVan** [comment]
@carlocaione @ioannisg @MaureenHelm @carlescufi I think the build failure are common for ARM64? So this PR is ready to be merged? Thanks.

File: boards/arm64/mimx8mm_evk_a53/mimx8mm_evk_a53.yaml [hunk 13]
```diff
@@ -0,0 +1,12 @@
+identifier: mimx8mm_evk_a53
+name: NXP i.MX8M Mini EVK A53
+type: mcu
+arch: arm
```
Comment (carlocaione) [inline boards/arm64/mimx8mm_evk_a53/mimx8mm_evk_a53.yaml:4]: The failure you see on CI are caused by this. It should be `arm64`.

**Review by carlocaione — CHANGES_REQUESTED** [review]
Let's wait for the CI to be all green.

File: boards/arm64/mimx8mm_evk_a53/mimx8mm_evk_a53.yaml (hunk 13 above)
Comment (MrVan) [reply boards/arm64/mimx8mm_evk_a53/mimx8mm_evk_a53.yaml:4]: oh. I missed to change this after aarch64 has a seperate folder.

**Comment by MrVan** [comment]
@carlocaione CI passed.

**Review by carlocaione — APPROVED** [review]

**Comment by MrVan** [comment]
@carlocaione @ioannisg @MaureenHelm @carlescufi Could this PR get merged now?

**Comment by carlocaione** [comment]
@MrVan we are in code freeze. This will be merged after release.

**Comment by MrVan** [comment]
@carlocaione ok, thanks for the info.

**Review by galak — DISMISSED** [review]
@MaureenHelm we need to figure out how to handle the drivers for this.  I don't think it makes sense to have duplicated HAL & non-HAL drivers.  Doubles our support efforts.

**Comment by MrVan** [comment]
> @MaureenHelm we need to figure out how to handle the drivers for this. I don't think it makes sense to have duplicated HAL & non-HAL drivers. Doubles our support efforts.

The effort to reuse NXP SDK https://github.com/NXPmicro/mcux-sdk/pull/11 has been rejected.

**Comment by carlocaione** [comment]
@MrVan what do you plan to do with this?

**Comment by MrVan** [comment]
> @MrVan what do you plan to do with this?

After https://github.com/zephyrproject-rtos/hal_nxp/pull/80 ready, I'll rework this PR.

**Comment by arvid-r** [comment]
Let me know if you need any testing related to this. I have the relevant boards (I believe) and would really like to see Jailhouse/Zephyr on arm64 at some point, so I'll try to help if I can.

**Comment by dleach02** [comment]
@MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.

**Comment by MrVan** [comment]
> @MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.

@dleach02 could your please share me the merged PR link?

**Comment by MrVan** [comment]
> > @MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.
> 
> @dleach02 could your please share me the merged PR link?
@dleach02
BTW: you mean i.MX8MM A53 port has been supported in NXP HAL?

**Comment by MrVan** [comment]
> > @MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.
> 
> @dleach02 could your please share me the merged PR link?
@dleach02
BTW: you mean i.MX8MM A53 port has been supported in NXP HAL?

**Comment by MrVan** [comment]
Update the PR, ram console is used here to avoid the dependency on nxp hal which block the PR for long time. After NXP HAL finally support ARM64, we could switch back. I am not sure why there is a twister build failure, it is caused this PR?

**Comment by dleach02** [comment]
@MrVan, Twister is showing failure in imx8mm_evk_jailhouse with samples/subsys/shell/shell_module/sample.shell.shell_module

**Comment by MrVan** [comment]
all check passed, please help check whether you are ok with this PR now. Thanks.

**Comment by dleach02** [comment]
@galak, can you revisit your change request to determine if the author has addressed your request?

**Comment by MrVan** [comment]
The major concern from @galak was the PR(previous version) has its own uart driver, not reusing NXP HAL. This conern has been addressed for quite some time, so any more comments from project owners?

File: dts/arm64/nxp/nxp_imx8mm_a53.dtsi [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+		#clock-cells = <3>;
+	};
+
+	uart2: serial@30890000 {
+		compatible = "nxp,imx-iuart";
+		reg = <0x30890000 DT_SIZE_K(64)>;
+		interrupts = <GIC_SPI 27 IRQ_TYPE_LEVEL 0>;
+		interrupt-names = "irq_0";
+		interrupt-parent = <&gic>;
+		clocks = <&ccm IMX_CCM_UART_CLK 0x6c 24>;
```
Comment (Mani-Sadhasivam) [inline dts/arm64/nxp/nxp_imx8mm_a53.dtsi:83]: The register offsets are wrong. Even in the M4 dts I did the same thing. Since the offsets were not used by the clock driver, these are irrelevant but we should fix the CCM binding to use `#clock-cells = <1>`.

This will also require changing a few dts, but that shouldn't be a big thing.

**Review by Mani-Sadhasivam — DISMISSED** [review]

File: soc/arm64/nxp_imx/mcimx8mm_a53/soc_a53.c [hunk 15]
```diff
@@ -16,7 +16,7 @@
  *
  * @return 0
  */
-static int mimx8mm_a53_init(struct device *arg)
+static int mimx8mm_a53_init(const struct device *arg)
```
Comment (Mani-Sadhasivam) [inline soc/arm64/nxp_imx/mcimx8mm_a53/soc_a53.c]: This change should be part of SoC commit.

File: dts/arm64/nxp/nxp_imx8mm_a53.dtsi (hunk 14 above)
Comment (MrVan) [reply dts/arm64/nxp/nxp_imx8mm_a53.dtsi:83]: If you think ccm binding need to change to `#clock-cells = <1>`, I could create the other PR to address it. But in this PR, I would not. I am not sure what 0x6c/24 means here, I just refer the m4 dtsi.

File: soc/arm64/nxp_imx/mcimx8mm_a53/soc_a53.c (hunk 15 above)
Comment (MrVan) [reply soc/arm64/nxp_imx/mcimx8mm_a53/soc_a53.c]: ok, I could squash the cleanup commit.

**Comment by dleach02** [comment]
@galak,

> @MaureenHelm we need to figure out how to handle the drivers for this. I don't think it makes sense to have duplicated HAL & non-HAL drivers. Doubles our support efforts.

Is this comment/block still valid?

File: dts/arm64/nxp/nxp_imx8mm_a53.dtsi (hunk 14 above)
Comment (Mani-Sadhasivam) [reply dts/arm64/nxp/nxp_imx8mm_a53.dtsi:83]: Okay. Let's handle it later.

**Comment by MrVan** [comment]
> @galak,
> 
> > @MaureenHelm we need to figure out how to handle the drivers for this. I don't think it makes sense to have duplicated HAL & non-HAL drivers. Doubles our support efforts.
> 
> Is this comment/block still valid?

Since no response. I am not sure people still happy to review this one year pending PR. I would create a new PR to restart the work.

**Comment by carlocaione** [comment]
@MrVan I don't think that's needed. Can you rebase? I'll try to push this.

File: soc/arm64/nxp_imx/imx8m/soc.h [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+#ifndef SOC_H
+#define SOC_H
+
+#include <sys/util.h>
+#include <toolchain.h>
+
+#ifndef _ASMLANGUAGE
+#include <devicetree.h>
+#endif
```
Comment (gmarull) [inline soc/arm64/nxp_imx/imx8m/soc.h:15]: this is not required, ref https://github.com/zephyrproject-rtos/zephyr/pull/41544

File: soc/arm64/nxp_imx/imx8m/soc_a53.c [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+ */
+static int mimx8mm_a53_init(const struct device *arg)
+{
+
+	/* TODO */
+
+	return 0;
+}
+
+SYS_INIT(mimx8mm_a53_init, PRE_KERNEL_1, 0);
```
Comment (gmarull) [inline soc/arm64/nxp_imx/imx8m/soc_a53.c:27]: can this be added when this does something?

File: boards/arm64/imx8mm_evk/dummy.c [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+			    NULL,					\
+			    NULL,					\
+			    NULL,					\
+			    PRE_KERNEL_1,				\
+			    CONFIG_SERIAL_INIT_PRIORITY,		\
+			    NULL);
+
+DT_INST_FOREACH_STATUS_OKAY(IUART_MCUX_INIT)
+
+#endif
```
Comment (gmarull) [inline boards/arm64/imx8mm_evk/dummy.c:35]: what's going on here?

**Review by gmarull — CHANGES_REQUESTED** [review]
A few questions

File: soc/arm64/nxp_imx/mcimx8mm_a53/soc_a53.c (hunk 15 above)
Comment (Mani-Sadhasivam) [reply soc/arm64/nxp_imx/mcimx8mm_a53/soc_a53.c]: please do so!

File: boards/arm64/imx8mm_evk/dummy.c (hunk 18 above)
Comment (MrVan) [reply boards/arm64/imx8mm_evk/dummy.c]: This file is not needed after rebase, dropped.

File: soc/arm64/nxp_imx/imx8m/soc_a53.c (hunk 17 above)
Comment (MrVan) [reply soc/arm64/nxp_imx/imx8m/soc_a53.c]: Dropped.

**Comment by MrVan** [comment]
@carlocaione @MaureenHelm  sorry for late reply, just back from vocation. I have addressed all the comments.

File: boards/arm64/imx8mm_evk/dummy.c (hunk 18 above)
Comment (MrVan) [reply boards/arm64/imx8mm_evk/dummy.c]: Seems CI not happy if this file dropped. Because UART is not enabled for now, so CI report.
samples/subsys/shell/shell_module/src/uart_reinit.c:19: undefined reference to `__device_dts_ord_9'

Bring it back

File: soc/arm64/nxp_imx/imx8m/soc.h [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+#include <sys/util.h>
+#include <toolchain.h>
+
+#ifndef _ASMLANGUAGE
+
+#include <device.h>
+
+#endif /* !_ASMLANGUAGE */
+
+/* TODO: Place Holder */
```
Comment (gmarull) [inline soc/arm64/nxp_imx/imx8m/soc.h:19]: none seem to be required, please, justify them. Maybe some other files are not including necessary headers?

File: boards/arm64/imx8mm_evk/imx8mm_evk_jailhouse.yaml [hunk 20]
```diff
@@ -0,0 +1,12 @@
+identifier: imx8mm_evk_jailhouse
+name: NXP i.MX8M Mini EVK A53 Jailhouse Inmate Cell
```
Comment (carlocaione) [inline boards/arm64/imx8mm_evk/imx8mm_evk_jailhouse.yaml:2]: can you add the info that is now in the commit message in the README?

File: soc/arm64/nxp_imx/imx8m/soc.h [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SOC_H
+#define SOC_H
+
+#include <sys/util.h>
+#include <toolchain.h>
+
+/* TODO: Place Holder */
```
Comment (gmarull) [inline soc/arm64/nxp_imx/imx8m/soc.h:13]: Why are these required? Likely some other files not including what they need?

Comment (MrVan) [reply soc/arm64/nxp_imx/imx8m/soc.h:13]: I just follow other arm64 platforms. I think soc.h is a central place for soc related header files, currently we have not add specific device support, so leave todo for now.

File: boards/arm64/imx8mm_evk/imx8mm_evk_jailhouse.yaml (hunk 20 above)
Comment (MrVan) [reply boards/arm64/imx8mm_evk/imx8mm_evk_jailhouse.yaml:2]: Add info to doc/index.rst.

File: soc/arm64/nxp_imx/imx8m/soc.h (hunk 21 above)
Comment (gmarull) [reply soc/arm64/nxp_imx/imx8m/soc.h:13]: > I just follow other arm64 platforms. I think soc.h is a central place for soc related header files, currently we have not add specific device support, so leave todo for now.

That doesn't mean others do it right, really. Ref https://github.com/zephyrproject-rtos/zephyr/issues/41543

Comment (MrVan) [reply soc/arm64/nxp_imx/imx8m/soc.h:13]: After a check, "arch/arm64/core/smp.c" includes soc.h, so could we refactor the header file later?

Comment (gmarull) [reply soc/arm64/nxp_imx/imx8m/soc.h:13]: Yes, fair enough. Could be part of #41543 work.

**Review by gmarull — APPROVED** [review]

**Comment by MrVan** [comment]
@carlocaione @MaureenHelm Ready for merge now?

**Comment by carlocaione** [comment]
> @carlocaione @MaureenHelm Ready for merge now?

Yes, but we are in stabilization period now. This will be merged when 3.0 will be released.

File: soc/arm64/nxp_imx/imx8m/mmu_regions.c [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+
+	MMU_REGION_FLAT_ENTRY("ANA_PLL",
+			      DT_REG_ADDR(DT_INST(0, nxp_imx_ana)),
+			      DT_REG_SIZE(DT_INST(0, nxp_imx_ana)),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_RW | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("UART",
+			      DT_REG_ADDR(DT_INST(1, nxp_imx_iuart)),
+			      DT_REG_SIZE(DT_INST(1, nxp_imx_iuart)),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_RW | MT_NS),
```
Comment (mbolivar-nordic) [inline soc/arm64/nxp_imx/imx8m/mmu_regions.c:41]: Hard-coded instance numbers like this are incorrect to use. Please use an alternate mechanism to specify exactly the nodes you want. Read the DT_INST docstring for more details.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: soc/arm64/nxp_imx/imx8m/mmu_regions.c (hunk 22 above)
Comment (MrVan) [reply soc/arm64/nxp_imx/imx8m/mmu_regions.c:41]: Could we use a follow up patch to address this? I just follow qemu platform here. @carlocaione  do you have any suggestions?

Comment (Mani-Sadhasivam) [reply soc/arm64/nxp_imx/imx8m/mmu_regions.c:41]: Considering the amount of time this PR has been in an open state (1 year), I think we should merge this and follow up with PRs for fixing issues like this. @mbolivar-nordic what do you think?

Comment (mbolivar-nordic) [reply soc/arm64/nxp_imx/imx8m/mmu_regions.c:41]: No, sorry, I don't want to see incorrect usage of the instance API merged.

Comment (MrVan) [reply soc/arm64/nxp_imx/imx8m/mmu_regions.c:41]: @mbolivar-nordic I am not experts in zephyr, so any suggestions what I should use?

**Comment by MrVan** [comment]
A bit werid, I have updated my 8mm-v1 branch, but this PR not update automatically.

**Comment by MrVan** [comment]
> A bit werid, I have updated my 8mm-v1 branch, but this PR not update automatically.

I am not sure why PR not update automatically. what could I do here? Close this PR and reopen a new one?

**Comment by stephanosio** [comment]
> > A bit werid, I have updated my 8mm-v1 branch, but this PR not update automatically.
> 
> I am not sure why PR not update automatically. what could I do here? Close this PR and reopen a new one?

@MrVan It looks like something broke on the GitHub side -- we have seen the same issue in other PRs lately. Please open a new PR.
